From 0d80ee00f7471317740f9843ca78ae96409377ee Mon Sep 17 00:00:00 2001
From: "Isaac J. Manjarres" <isaacmanjarres@google.com>
Date: Tue, 19 May 2026 18:51:25 +0000
Subject: [PATCH 155/223] ANDROID: kasan: Don't instrument Rust globals if
 UNWIND_PATCH_PAC_INTO_SCS=y

When the generic variant of KASAN is used, constructors are generated to
setup KASAN's redzones for global variables. However, the following
conditions have been observed to lead to memory corruption:

1. CONFIG_RUST=y
2. CONFIG_SHADOW_CALL_STACK=y
3. CONFIG_UNWIND_PATCH_PAC_INTO_SCS=y
4. The CPUs on the device do not support the pointer authentication code
   (PAC) feature.

The last condition results in the kernel code being patched to use
the shadow call stack feature instead. This causes the kernel to
patch the PAC instructions to their shadow call stack push and pop
instructions during boot. However, this is not behaving as expected
at the moment and manifests as memory corruption.

This has only been observed with the rust stdlib module, but could
happen with other modules. Therefore, disable instrumentation of Rust
global variables if CONFIG_UNWIND_PATCH_PAC_INTO_SCS is enabled until
this has been fixed properly.

Bug: 513032835
Bug: 511009778
Change-Id: I474a87f8c765c96e3a83eb53c08cfd04986f8aa3
Signed-off-by: Isaac J. Manjarres <isaacmanjarres@google.com>
---
 scripts/Makefile.kasan | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/scripts/Makefile.kasan b/scripts/Makefile.kasan
index 91504e81247a..5091ffe4d945 100644
--- a/scripts/Makefile.kasan
+++ b/scripts/Makefile.kasan
@@ -100,7 +100,31 @@ CFLAGS_KASAN += $(call check-args, cc-param, $(kasan_params))
 
 ifdef CONFIG_RUST
 	# Avoid calling `rustc-param` unless Rust is enabled.
-	RUSTFLAGS_KASAN += $(call check-args, rustc-param, $(kasan_params))
+	#
+	# b/513032835: TODO: Re-enable instrumentation for globals once it's working properly.
+	#
+	# When CONFIG_KASAN_GENERIC is enabled, constructors are generated to setup KASAN's redzones
+	# for global variables. However, under the following conditions, memory corruption has been
+	# observed:
+	#
+	# 1. CONFIG_RUST=y
+	# 2. CONFIG_SHADOW_CALL_STACK=y
+	# 3. CONFIG_UNWIND_PATCH_PAC_INTO_SCS=y
+	# 4. The CPUs on the device do not support the pointer authentication code (PAC) feature
+	#
+	# Condition 4 causes the kernel to fallback to using the shadow call stack feature by
+	# patching the PAC instructions with loads and stores from and to the shadow call stack
+	# respectively. However, this is not happening as expected at the moment and manifests
+	# as memory corruption.
+	#
+	# While this has only been observed with the Rust stdlib module (rust/core.o), it could
+	# happen with other Rust modules. So, disable instrumentation of Rust globals if needed.
+ifdef CONFIG_UNWIND_PATCH_PAC_INTO_SCS
+	rust_kasan_params := $(subst asan-globals=1,asan-globals=0,$(kasan_params))
+else
+	rust_kasan_params := $(kasan_params)
+endif
+	RUSTFLAGS_KASAN += $(call check-args, rustc-param, $(rust_kasan_params))
 endif # CONFIG_RUST
 
 export CFLAGS_KASAN CFLAGS_KASAN_NOSANITIZE RUSTFLAGS_KASAN
-- 
2.50.1

